backend/tracim_backend/lib/search/search.py

```diff
@@ -0,0 +1,430 @@
+            show_active=show_active,
+        )
+        total_hits = 0
+        keywords = content_api.get_keywords(search_string)
+        offset = self.offset_from_pagination(size, page_nb)
+        content_list, total_hits = content_api.search(
+            keywords=keywords, size=size, offset=offset, content_types=content_types
+        )
+        content_in_context_list = []
+        for content in content_list:
+            content_in_context_list.append(content_api.get_content_in_context(content))
+        return SimpleContentSearchResponse(content_in_context_list, total_hits)
+
+
+class ESSearchApi(SearchApi):
+    """
+    Search using ElasticSearch :
+    - need indexing content first
+    - allow pagination and filtering by content_type, deleted, archived
+    - support ranking
+    - search in content file for html-doc and thread
+    - search in content file for file if ingest mode activated
+    """
+
+    def __init__(self, session: Session, current_user: typing.Optional[User], config: CFG) -> None:
+        super().__init__(session, current_user, config)
+        assert config.SEARCH__ENABLED
+        assert config.SEARCH__ENGINE == ELASTICSEARCH__SEARCH_ENGINE_SLUG
+        # TODO - G.M - 2019-05-31 - we do support only "one elasticsearch server case here in config,
+        # check how to support more complex case.
+        self.es = Elasticsearch(
+            hosts=[
+                (
+                    {
+                        "host": self._config.SEARCH__ELASTICSEARCH__HOST,
+                        "port": self._config.SEARCH__ELASTICSEARCH__PORT,
+                    }
+                )
+            ]
+        )
+
+    def create_index(self) -> None:
+        """
+        Create the index template in elasticsearch specifying the mappings and any
+        settings to be used. This can be run at any time, ideally at every new code
+        deploy
+        """
+        # INFO - G.M - 2019-05-15 - alias migration mecanism to allow easily updatable index.
+        # from https://github.com/elastic/elasticsearch-dsl-py/blob/master/examples/alias_migration.py
+        # Configure index with our indexing preferences
+        logger.info(self, "Create index settings ...")
+        if self._config.SEARCH__ELASTICSEARCH__USE_INGEST:
+            self._create_ingest_pipeline()
+        # create an index template
+        index_template = IndexedContent._index.as_template(
+            INDEX_DOCUMENTS_ALIAS, INDEX_DOCUMENTS_PATTERN
+        )
+        # upload the template into elasticsearch
+        # potentially overriding the one already there
+        index_template.save(using=self.es)
+
+        # create the first index if it doesn't exist
+        if not IndexedContent._index.exists(using=self.es):
+            self.migrate_index(move_data=False)
+
+        logger.info(self, "ES index is ready")
+
+    def refresh_index(self) -> None:
+        self.es.indices.refresh(INDEX_DOCUMENTS_ALIAS)
```

Comment:
Can you add a docstring about what elastic search do here ?

Reply:
done.